Sources/OpenCombine/Publishers/Publishers.Filter.swift

```diff
@@ -0,0 +1,213 @@
+    /// Boolean value indicating whether to republish the element.
+    /// - Returns:  A publisher that republishes all elements that satisfy the closure.
+    public func tryFilter(
+        _ isIncluded: @escaping (Self.Output) throws -> Bool
+    ) -> Publishers.TryFilter<Self> {
+        return Publishers.TryFilter(upstream: self, isIncluded: isIncluded)
+    }
+}
+
+private final class _Filter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+      Subscriber,
+      CustomStringConvertible
+    where Upstream.Output == Downstream.Input,
+          Upstream.Failure == Downstream.Failure {
+
+    var description: String { return "Filter" }
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 },
+                   isIncluded: { .success(isIncluded($0))
+        })
+    }
+}
+
+private final class _TryFilter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+    Subscriber,
+    CustomStringConvertible
+    where Upstream.Output == Downstream.Input, Downstream.Failure == Error {
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) throws -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 as Error },
+                   isIncluded: { input in Result { try isIncluded(input) }
+        })
+    }
+
+    var description: String { return "TryFilter" }
+}
+
+private class _BaseFilter<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Output = Downstream.Input
+    typealias Failure = Upstream.Failure
+
+    private let _isIncluded: (Input) -> Result<Bool, Downstream.Failure>
+    private var _unsatisfiedDemand: Subscribers.Demand = .none
+    private var _isFinished = false
+
+    var errorTransform: (Upstream.Failure) -> Downstream.Failure
+
+    init(downstream: Downstream,
+         errorTransform: @escaping (Upstream.Failure) -> Downstream.Failure,
+         isIncluded: @escaping (Input) -> Result<Bool, Downstream.Failure>) {
+        self._isIncluded = isIncluded
+        self.errorTransform = errorTransform
+        super.init(downstream: downstream)
+    }
+
+    func receive(subscription: Subscription) {
+        upstreamSubscription = subscription
+        downstream.receive(subscription: self)
+    }
+
+    func receive(_ input: Input) -> Subscribers.Demand {
+        switch _isIncluded(input) {
+        case .failure(let error):
+            downstream.receive(completion: .failure(error))
+            cancel()
+            _isFinished = true
+            return .none
+        case .success(let isIncluded):
+            if isIncluded {
+                let newDemand = downstream.receive(input)
+                _unsatisfiedDemand += newDemand - 1
```

Comment:
If `newDemand` is `.max(0)`, this will crash as subtracting from zero demand is not permitted.

Better use two in-place mutations here:

```swift
_unsatisfiedDemand -= 1
_unsatisfiedDemand += newDemand
```